### PR TITLE
Fix calldata decoding overflow bug

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Review by ekpyron — COMMENTED** [review]
Looks good, but needs at least a changelog entry.
Also: shouldn't this affect some of the gas cost tests - it doesn't seem so, but isn't that a bit weird?

**Comment by ekpyron** [comment]
Also I do wonder, whether we should unify the behaviour between memory and calldata decoding at some point... should we create an issue for that at least?

File: docs/bugs.json [hunk 1]
```diff
@@ -1,4 +1,15 @@
 [
+    {
+        "name": "ABIDecodeTwoDimensionalArray",
```
Comment (ekpyron) [inline docs/bugs.json]:
```suggestion
        "name": "ABIDecodeTwoDimensionalArrayMemory",
```
?

File: test/libsolidity/gasTests/abiv2_optimised.sol [hunk 2]
```diff
@@ -17,9 +17,9 @@ contract C {
 // optimize-yul: true
 // ----
 // creation:
-//   codeDepositCost: 626600
-//   executionCost: 657
-//   totalCost: 627257
+//   codeDepositCost: 664600
```
Comment (ekpyron) [inline test/libsolidity/gasTests/abiv2_optimised.sol:20]: Quite a difference... does this include the additional debug reason string?

**Review by ekpyron — DISMISSED** [review]
I would have chosen a more specific bug identifier, but the code change looks good to me in any case!

File: test/libsolidity/gasTests/abiv2_optimised.sol (hunk 2 above)
Comment (chriseth) [reply test/libsolidity/gasTests/abiv2_optimised.sol:20]: I checked, it includes 10 new code snippets. The whole contract is essentially only an ABI decoder, so i think this is reasonable.

**Comment by chriseth** [comment]
Updated.

**Review by ekpyron — APPROVED** [review]

**Pull request merged**